docs/spec_draft.md

```diff
@@ -3731,6 +3808,80 @@ where `pred_val = rank(pred) == 0 ? pred : pred[i0, ..., iR-1]`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.send
+
+### Semantics
+
+Sends data `inputs` from one partition to a [stablehlo.recv](#stablehlorecv)
+instruction, in another partition, that shares the same channel handle pair
+{`channel_id`, `channel_type`}.
+
+`stablehlo.send`, paired with the corresponding [stablehlo.recv](#stablehlorecv)
+instruction, represents synchronous communication. However, in order to enable
+asynchronous data transfers, it is decomposed into two HLO instructions
+(`Send` and `SendDone`). Similarly, the paired
+[stablehlo.recv](#stablehlorecv) is internally decomposed into two HLO
+instructions (`Recv` and `RecvDone`).  [todo](https://github.com/openxla/stablehlo/issues/642)
+
+These 4 instructions, each associated with a particular channel, has the
+following semantics:
+
+  * **Recv**: Initiator of the channel by allocating the receive buffer and
+              notifying the sender that it is ready to receive data.
+  * **Send**: Waits for the Recv notification and starts the data transfer to
+              the receive buffer.
+  * **SendDone**: Waits until the Send operation is done, so that the sender
+                  buffer can be reused for other purposes.
+  * **RecvDone**: waits until the data transfer is done and the receive buffer
+                  is ready for use.
+
+The diagram shows the execution order of the instructions.
+
+<img align="center" src="spec_draft/happens-before.svg" />
+
+The `channel_type` type has the following semantics:
+
+  * `DEVICE_TO_DEVICE`: A channel for sending data between devices.
+  * `DEVICE_TO_HOST`: A channel for sending data from the device to the host.
+                      Can only be used with a `stablehlo.send` operation.
+  * `HOST_TO_DEVICE`: A channel for sending data from the host to the device.
+                      Can only be used with a `stablehlo.recv` operation.
+
+If `is_host_transfer`, an optional input with default value as `fasle`, is
+`true`, then the operation transfers data to the host.
+
+### Inputs
+
+| Name               | Type                                                               |
+|--------------------|--------------------------------------------------------------------|
+| `inputs`           | variadic number of tensors of any supported type                   |
+| `token`            | `token`                                                            |
+| `channel_id`       | constant of type `si64`                                            |
+| `channel_type`     | enum of `DEVICE_TO_DEVICE`, `DEVICE_TO_HOST`, and `HOST_TO_DEVICE` |
+| `is_host_transfer` | constant of type `i1`                                              |
+
+### Outputs
+
+| Name      | Type  |
+|-----------|-------|
+| `results` | token |
+
+### Constraints
+  * (C1) [todo](https://github.com/openxla/stablehlo/issues/579) `channel_type` must be
```

Comment:
size of inputs > 0 ?

Reply:
Thanks!

Based on https://github.com/tensorflow/tensorflow/blob/26ad1612bf16002c845fef4c56758784e804dcad/tensorflow/compiler/xla/translate/mhlo_to_hlo/mlir_hlo_to_hlo.cc#L1984

the size of inputs can be [zero](https://github.com/tensorflow/tensorflow/blob/26ad1612bf16002c845fef4c56758784e804dcad/tensorflow/compiler/xla/client/xla_builder.cc#L3675) or [not](https://github.com/tensorflow/tensorflow/blob/26ad1612bf16002c845fef4c56758784e804dcad/tensorflow/compiler/xla/client/xla_builder.cc#L3743). 
Same for Recv.  
 

Updated the text.